pkg/template/template.go

```diff
@@ -19,37 +20,621 @@ package template
+// templated like above as:
+//
+// val: abc
+func noop(given ...interface{}) (op string) {
+	return
+}
+
+// pickSuffix picks and returns an item from an array of strings. It picks the
+// first item whose suffix matches the provided match.
+//
+// NOTE:
+//  This is intended to be used as a template function
+//
+// Example:
+//  {{- "jiva-rep, cstor" | splitList ", " | pickSuffix "-rep" | noop -}}
+func pickSuffix(match string, given []string) (matched string) {
+	for _, givenItem := range given {
+		if strings.HasSuffix(givenItem, match) {
+			matched = givenItem
+			return
+		}
+	}
+
+	return
+}
+
+// pickPrefix picks and returns an item from an array of strings. It picks the
+// first item whose prefix matches the provided match.
+//
+// NOTE:
+//  This is intended to be used as a template function
+//
+// Example:
+//  {{- "pvc-jiva-rep, cstor" | splitList ", " | pickPrefix "pvc-" | noop -}}
+func pickPrefix(match string, given []string) (matched string) {
+	for _, givenItem := range given {
+		if strings.HasPrefix(givenItem, match) {
+			matched = givenItem
+			return
+		}
+	}
+
+	return
+}
+
+// pickContains picks and returns an item from an array of strings. It picks the
+// first item which contains the provided match.
+//
+// NOTE:
+//  This is intended to be used as a template function
+//
+// Example:
+//  {{- "pvc-jiva-rep, cstor" | splitList ", " | pickContains "-jiva-" | noop -}}
+func pickContains(match string, given []string) (matched string) {
+	for _, givenItem := range given {
+		if strings.Contains(givenItem, match) {
+			matched = givenItem
+			return
+		}
+	}
+
+	return
+}
+
+// splitKeyMap builds a list of map of key:value pairs where each map is set
+// against a key. It accepts an array where each array item is converted into a
+// map. Each map is set against a key known as primary key i.e. pkey. The
+// resulting pkey:map pairs are set against the destination object. This
+// function is same as 'keyMap' with user defined splitters as extra
+// parameters here.
+//
+// NOTE:
+//  Technically, it forms a map[string]interface{} structure where interface{}
+// is a map[string]string.
+//
+// NOTE:
+//  This is intended to be used as a template function
 //
 // NOTE:
-//  It will always return a string, even on error.
-//  This is designed to be called from a template.
+//  This appends current value to original value if any.
+//
+// Example:
+//  {{- "pkey=openebs--stor1=jiva--stor2=cstor" | splitList " " | splitKeyMap "-- =" "vals" .Target | noop -}}
+//  {{- "co1=swarm--co2=k8s" | splitList " " | splitKeyMap "-- =" "vals" .Target | noop -}}
+//  {{- "pkey=openebs--stor2=mstor" | splitList " " | splitKeyMap "-- =" "vals" .Target | noop -}}
+//
+// Above will result into following:
+//  Target: map[string]interface{}{
+//    "vals": map[string]interface{}{
+//      "openebs": map[string]interface{}{
+//        "stor1": "jiva",
+//        "stor2": "cstor, mstor",
+//      },
+//      "pkey": map[string]interface{}{
```

Comment:
Is the example correct? should it be something else and not "pkey" in this line?

Reply:
Example is correct.

Reply:
this is correct. if there is no primary key in a given string i.e. a k:v pair with k as `pkey` then this function will default the primary key as "pkey"